OpenRA.Mods.Common/AudioLoaders/Mp3Loader.cs

```diff
@@ -0,0 +1,60 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.IO;
+using MP3Sharp;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Common.AudioLoaders
+{
+	public class Mp3Loader : ISoundLoader
+	{
+		bool ISoundLoader.TryParseSound(Stream stream, out ISoundFormat sound)
+		{
+			try
+			{
+				sound = new Mp3Format(stream);
+				return true;
+			}
+			catch
+			{
+				// Not a (supported) MP3
+			}
+
+			sound = null;
+			return false;
+		}
+	}
+
+	public sealed class Mp3Format : ISoundFormat
+	{
+		public int Channels { get { return mp3.ChannelCount; } }
+		public int SampleBits { get { return 16; } }
+		public int SampleRate { get { return mp3.Frequency; } }
+		public float LengthInSeconds { get { return mp3.Length / mp3.Frequency; } }
```

Comment:
It looks like `mp3.Length` returns the length of the stream in bytes. Since MP3 is compressed, this is not going to have any direct relationship to the audio length in seconds.

Reply:
If this is actually the length of the uncompressed data then you will need to account for the channel count and sample bits when converting to time.

Reply:
It is an estimation https://chunminchang.github.io/blog/post/estimation-of-mp3-duration

Reply:
I tweaked this a bit to give better results with the example files I used. There is in general no way of knowing how long the duration is until it is completely parsed which I guess we want to avoid. Secondly the MP3Sharp library does not expose the total amount of samples or even the sample bits.